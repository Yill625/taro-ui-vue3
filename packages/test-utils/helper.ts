import { mount, shallowMount, DOMWrapper, VueWrapper } from '@vue/test-utils'
import { VNode, Slot } from '@vue/runtime-core'

export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))

export interface Slots {
  default: Array<string | VNode> | Slot
  [x: string]: Array<string | VNode> | Slot
}

export function mountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return mount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export function genMountFn(
  comp: any,
  components?: Record<string, object>,
  useShallowMount?: boolean
) {
  return (props = {}, slots?: Slots) => {
    return useShallowMount
      ? shallowMountFactory(comp, components, props, slots)
      : mountFactory(comp, components, props, slots)
  }
}

export function shallowMountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return shallowMount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export interface Dimension {
  top?: number
  bottom?: number
  left?: number
  right?: number
  width?: number
  height?: number
}

export function genDelayedSelectorSpy(
  utils: any,
  dimensions: Dimension
): jest.SpyInstance {
  return jest
    .spyOn(utils, 'delayQuerySelector')
    .mockImplementation((_, sel: string, delay?: number) => {
      return new Promise((resolve) => {
        resolve([dimensions])
      })
    })
}

export async function triggerTouchEvents(
  el: DOMWrapper<Element>,
  touchstartDetails: any,
  touchendDetails: any
) {
  const startDetails = {
    touches: [touchstartDetails],
    changedTouches: [touchstartDetails]
  }

  const endDetails = {
    touches: [touchendDetails],
    changedTouches: [touchendDetails]
  }

  await el.trigger('touchstart', startDetails)
  await el.trigger('touchmove', endDetails)
  await el.trigger('touchend', endDetails)
}

export function testPropClassAndStyle(
  mountFn: (props?: {}, slots?: Slots | undefined) => VueWrapper<any>
) {

  it.concurrent('should render prop class', async () => {
    const wrapper = mountFn({ class: 'test' })
    expect(
      wrapper.classes()
    ).toContain('test')
  })

  it.concurrent('should render prop style', async () => {
    const wrapper = mountFn({ style: { color: 'red' } })
    expect(
      wrapper.attributes('style')
    ).toContain('color: red;')
  })
}

export function hexToRGBA(hex: string, opacity = 1) {
  hex = hex.replace('#', '')
  const rgb = hex
    .match(new RegExp('(.{' + hex.length / 3 + '})', 'g'))!
    .map((l) => parseInt(hex.length % 2 ? l + l : l, 16))

  const rgba = rgb.concat(isFinite(opacity) ? opacity : 1)

  return opacity === 1
    ? `rgb(${rgb.join(', ')})`
    : `rgba(${rgba.join(', ')})`
}