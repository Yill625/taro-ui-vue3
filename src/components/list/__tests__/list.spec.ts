import { mountFactory } from '@/tests/helper'
import { ref } from 'vue'
import AtList from '../index'
import AtListItem from '../item'

describe('AtList', () => {
  it('should render all nodes and match snapshot', () => {
    const wrapper = mountFactory(AtList, {}, {}, {
      default: ['slot content']
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render default slot', () => {
    const wrapper = mountFactory(AtList, {}, {}, {
      default: ['slot content']
    })

    expect(
      wrapper
        .find('.at-list')
        .text()
    ).toBe('slot content')
  })

  it('should render prop class', () => {
    const wrapper = mountFactory(AtList, {}, { class: 'test' })
    expect(
      wrapper
        .find('.at-list')
        .classes('test')
    ).toBeTruthy()
  })

  it('should render prop style', () => {
    const wrapper = mountFactory(AtList, {}, { style: 'color: red;' })
    expect(
      wrapper
        .find('.at-list')
        .attributes('style')
    ).toBe('color: red;')
  })

  it('should render prop hasBorder', async () => {
    const wrapper = mountFactory(AtList)
    expect(
      wrapper
        .find('.at-list--no-border')
        .exists()
    ).toBeFalsy()

    await wrapper.setProps({ hasBorder: false })
    expect(
      wrapper
        .find('.at-list--no-border')
        .exists()
    ).toBeTruthy()
  })
})

const iconInfo = {
  size: 25,
  color: '#78A4FA',
  value: 'calendar',
}

describe('AtListItem', () => {
  it('should render all nodes and match snapshot', () => {
    const wrapper = mountFactory(AtListItem, {}, {
      class: 'test',
      style: 'color: red;',
      note: 'note',
      title: 'title',
      thumb: 'http://img.test.com/3N31e.png',
      extraText: 'extraText',
      arrow: 'right',
      iconInfo: { value: 'calendar' }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render prop class', () => {
    const wrapper = mountFactory(AtListItem, {}, { class: 'test' })
    expect(
      wrapper
        .find('.at-list__item')
        .classes('test')
    ).toBeTruthy()
  })

  it('should render prop style', () => {
    const wrapper = mountFactory(AtListItem, {}, { style: 'color: red;' })
    expect(
      wrapper
        .find('.at-list__item')
        .attributes('style')
    ).toBe('color: red;')
  })

  it('should render prop thumb', () => {
    const wrapper = mountFactory(AtListItem, {}, {
      thumb: 'http://img.test.com/3N31e.png'
    })

    const thumbEl = wrapper.find('.at-list__item-thumb')
    expect(thumbEl.exists()).toBeTruthy()
    expect(thumbEl.element).toMatchSnapshot()

    expect(
      wrapper
        .find('.at-list__item-thumb > .item-thumb__info')
        .attributes('src')
    ).toBe('http://img.test.com/3N31e.png')

  })

  it('should render prop iconInfo', () => {
    const wrapper = mountFactory(AtListItem, {}, {
      iconInfo: {
        ...iconInfo,
        prefixClass: 'prefixClass',
        class: 'klass',
        style: 'background: teal;'
      }
    })

    const iconWrapperEl = wrapper.find('.at-list__item-icon')

    expect(iconWrapperEl.exists()).toBeTruthy()
    expect(iconWrapperEl.element).toMatchSnapshot()

    const iconEl = wrapper.find('.at-list__item-icon > view')

    expect(iconEl.classes()).toEqual([
      'prefixClass',
      'prefixClass-calendar',
      'klass'
    ])

    expect(iconEl.attributes('style')).toEqual(
      'color: rgb(120, 164, 250); font-size: 25px; background: teal;'
    )
  })

  it.each([
    'title',
    'note'
  ])('should render prop %s', (propName) => {
    const wrapper = mountFactory(AtListItem, {}, {
      [propName]: propName
    })
    expect(
      wrapper
        .find(`.item-content__info-${propName}`)
        .text()
    ).toBe(propName)
  })

  it('should render prop extraText and match snapshot', () => {
    const wrapper = mountFactory(AtListItem, {}, {
      extraText: 'extraText'
    })

    const extraTextEl = wrapper.find(`.item-extra__info`)
    expect(extraTextEl.text()).toBe('extraText')
    expect(extraTextEl.element).toMatchSnapshot()
  })

  it('should render prop extraThumb and match snapshot', () => {
    const wrapper = mountFactory(AtListItem, {}, {
      extraThumb: 'http://img.test.com/3N31e.png'
    })
    expect(
      wrapper
        .find(`.item-extra__image-info`)
        .attributes('src')
    ).toBe('http://img.test.com/3N31e.png')

    expect(
      wrapper
        .find(`.item-extra__image`)
        .element
    ).toMatchSnapshot()
  })

  it('should render prop isSwitch and match snapshot', async () => {
    const wrapper = mountFactory(AtListItem, {}, {
      isSwitch: true,
      switchColor: 'red'
    })

    const switchWrapperEl = wrapper.find('.item-extra__switch')
    expect(switchWrapperEl.exists()).toBeTruthy()
    expect(switchWrapperEl.element).toMatchSnapshot()

    const swithEl = wrapper.find('.item-extra__switch > switch')

    expect(
      swithEl.attributes('checked')
    ).toBe("false")

    expect(
      swithEl.attributes('disabled')
    ).toBe('false')

    expect(
      swithEl.attributes('color')
    ).toBe('red')

    await wrapper.setProps({ switchIsCheck: true })
    expect(
      wrapper
        .find('.item-extra__switch > switch')
        .attributes('checked')
    ).toBe("true")
  })

  it.each([
    'up', 'down', 'right'
  ])('should render prop arrow -- %s', (option) => {
    const wrapper = mountFactory(AtListItem, {}, {
      arrow: option
    })

    const extraIconWrapperEl = wrapper.find('.item-extra__icon')
    expect(extraIconWrapperEl.exists()).toBeTruthy()

    expect(
      wrapper
        .find('.item-extra__icon-arrow')
        .classes()
    ).toContain(`at-icon-chevron-${option}`)
  })
})

describe('AtListItem Behavior ', () => {
  it('should trigger onClick', async () => {
    const onClick = jest.fn()
    const wrapper = mountFactory(AtListItem, {}, {
      onClick
    })

    await wrapper.find('.at-list__item').trigger('tap')
    expect(onClick).toBeCalled()
  })

  it('should trigger onSwitchChange', async () => {
    const checked = ref(false)
    const onSwitchChange = jest.fn((e) => {
      checked.value = e.detail
    })

    const wrapper = mountFactory(AtListItem, {}, {
      isSwitch: true,
      switchIsCheck: checked.value,
      onSwitchChange
    })

    await wrapper
      .find('.item-extra__switch > switch')
      .trigger('change', { detail: true })

    expect(onSwitchChange).toBeCalled()
    expect(checked.value).toBe(true)
  })

  it('should not trigger any event when disabled', async () => {
    const onSwitchChange = jest.fn()
    const onClick = jest.fn()

    const wrapper = mountFactory(AtListItem, {}, {
      disabled: true,
      isSwitch: true,
      onClick,
      onSwitchChange
    })

    await wrapper.find('.at-list__item').trigger('tap')
    await wrapper
      .find('.item-extra__switch > switch')
      .trigger('change', { detail: true })

    expect(onClick).not.toBeCalled()
    expect(onSwitchChange).not.toBeCalled()
  })



  // @TODO: check how to mock event.stopProparation()
  it.skip('should stop propagation when clicking switch wrapper', async () => {
    const ev = { stopPropagation: jest.fn() }

    const wrapper = mountFactory(AtListItem, {}, {
      isSwitch: true
    })

    await wrapper
      .find('.item-extra__switch')
      .trigger('tap', ev)

    expect(ev.stopPropagation).toBeCalled()
  })
})
