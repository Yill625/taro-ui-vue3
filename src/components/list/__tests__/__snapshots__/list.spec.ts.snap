// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtList should render all nodes and match snapshot 1`] = `
<view
  class="at-list"
>
  slot content
</view>
`;

exports[`AtListItem should render all nodes and match snapshot 1`] = `
<view
  class="test at-list__item at-list__item--thumb at-list__item--multiple at-list__item--no-border test"
  style="color: red;"
>
  <view
    class="at-list__item-container"
  >
    <view
      class="at-list__item-thumb item-thumb"
    >
      <image
        class="item-thumb__info"
        mode="scaleToFill"
        src="http://img.test.com/3N31e.png"
      />
    </view>
    <view
      class="at-list__item-icon item-icon"
    >
      <view
        class="at-icon at-icon-calendar"
        style=""
      />
    </view>
    <view
      class="at-list__item-content item-content"
    >
      <view
        class="item-content__info"
      >
        <view
          class="item-content__info-title"
        >
          title
        </view>
        <view
          class="item-content__info-note"
        >
          note
        </view>
      </view>
    </view>
    <view
      class="at-list__item-extra item-extra"
    >
      <view
        class="item-extra__info"
      >
        extraText
      </view>
      
      <!---->
      <view
        class="item-extra__icon"
      >
        <view
          class="at-icon item-extra__icon-arrow at-icon-chevron-right"
        />
      </view>
    </view>
  </view>
</view>
`;

exports[`AtListItem should render prop extraText and match snapshot 1`] = `
<view
  class="item-extra__info"
>
  extraText
</view>
`;

exports[`AtListItem should render prop extraThumb and match snapshot 1`] = `
<view
  class="item-extra__image"
>
  <image
    class="item-extra__image-info"
    mode="aspectFit"
    src="http://img.test.com/3N31e.png"
  />
</view>
`;

exports[`AtListItem should render prop iconInfo 1`] = `
<view
  class="at-list__item-icon item-icon"
>
  <view
    class="prefixClass prefixClass-calendar klass"
    style="color: rgb(120, 164, 250); font-size: 25px; background: teal;"
  />
</view>
`;

exports[`AtListItem should render prop isSwitch and match snapshot 1`] = `
<view
  class="item-extra__switch"
>
  <switch
    checked="false"
    color="red"
    disabled="false"
  />
</view>
`;

exports[`AtListItem should render prop thumb 1`] = `
<view
  class="at-list__item-thumb item-thumb"
>
  <image
    class="item-thumb__info"
    mode="scaleToFill"
    src="http://img.test.com/3N31e.png"
  />
</view>
`;
